lib/ocpp/v201/database_handler.cpp

```diff
@@ -733,26 +743,213 @@ void DatabaseHandler::insert_or_update_charging_profile(const int evse_id, const
     stmt->bind_int("@evse_id", evse_id);
     stmt->bind_int("@stack_level", profile.stackLevel);
     stmt->bind_text("@charging_profile_purpose",
-                    conversions::charging_profile_purpose_enum_to_string(profile.chargingProfilePurpose));
+                    conversions::charging_profile_purpose_enum_to_string(profile.chargingProfilePurpose),
+                    SQLiteString::Transient);
+    if (profile.transactionId.has_value()) {
+        stmt->bind_text("@transaction_id", profile.transactionId.value().get(), SQLiteString::Transient);
+    } else {
+        stmt->bind_null("@transaction_id");
+    }
+
     stmt->bind_text("@profile", json_profile.dump(), SQLiteString::Transient);
+    stmt->bind_text("@charging_limit_source", conversions::charging_limit_source_enum_to_string(charging_limit_source));
 
     if (stmt->step() != SQLITE_DONE) {
         throw QueryExecutionException(this->database->get_error_message());
     }
 }
 
-void DatabaseHandler::delete_charging_profile(const int profile_id) {
+bool DatabaseHandler::delete_charging_profile(const int profile_id) {
     std::string sql = "DELETE FROM CHARGING_PROFILES WHERE ID = @profile_id;";
     auto stmt = this->database->new_statement(sql);
 
     stmt->bind_int("@profile_id", profile_id);
     if (stmt->step() != SQLITE_DONE) {
         throw QueryExecutionException(this->database->get_error_message());
     }
+
+    return stmt->changes() > 0;
+}
+
+bool DatabaseHandler::clear_charging_profiles() {
+    return this->database->clear_table("CHARGING_PROFILES");
+}
+
+bool DatabaseHandler::clear_charging_profiles_matching_criteria(const std::optional<int32_t> profile_id,
```

Comment:
The criteria for `get_charging_profiles_matching_criteria ()` and `clear_charging_profiles_matching_criteria()` are differnt types with different requirements.

Reply:
Can you articulate what these differences are?

For example:
- I look at the implementation of `clear_charging_profiles_matching_criteria` and it checks for evse_id, stack level, chargingProfilePurpose....
- I look at the implementation of `get_charging_profiles_matching_criteria` and it checks for evse_id, charging_profile_id, charging_profile_purpose,...

Maybe this is just a refactor to an internal method, but it seems like we should be able to avoid repeating checks. I am willing to make this be a non-blocking comment, but given that this is already a cleanup change, want to see if we can just handle this now...

Reply:
`get_charging_profiles_matching_criteria()` takes in an optional `evse_id` and `ChargingProfileCriterion` type. `ChargingProfileCriterion` expects a list of profile IDs, the charging limit source, the purpose, and the stackLevel. The function's return type is also very specific to the use case of K09, returning a tuple of the evse_id, profile, and the source.

`clear_charging_profiles_matching_criteria()` takes in an optional `profile_id` and an optional `ClearChargingProfile` type. This type expects an optional evse ID, purpose, and stackLevel. We would need to convert the types we receive in here to what `get_charging_profiles_matching_criteria()` expects.

We do not need the extra information that `get` provides in clear, we just want to delete the matching profiles.

In addition, one of the goals of this change set was to avoid iteration and extra queries where possible. This proposed change would mean allocating a list of profiles unnecessarily, iterating over it, and then for each profile creating a new delete query instead of deleting all profiles in one query without iteration.

Reply:
The first argument is not very convincing to me, but the second

> In addition, one of the goals of this change set was to avoid iteration and extra queries where possible. This proposed change would mean allocating a list of profiles unnecessarily, iterating over it, and then for each profile creating a new delete query instead of deleting all profiles in one query without iteration.

is a pretty good argument

LGTM!